### ISSUE TITLE
Steal a bikehorn

### ISSUE THREAD

**Issue by Profakos** [SuperSayu/tgstation#98]
Because we iterate though every steal item types, instead of just the 'steal' subtypes (presumable so ninja objective giving does not need a seperate proc). As a result, "Steal a bikehorn" was included in someone's objectives, which is the base objective type.

**Comment by SuperSayu** [comment]
There may also be other objective types that have no details and are therefore the basic objective, I thought I had ripped out most of them but anything that is not given new details could be it.  It would be nice to have a log of what the objective's type actually was, that would make it easy to ferret out, but that's more complicated to get and I don't really expect you to find out.

If you do see it again and want to find out, debug/view vars the mob, look in their mind (variable's name is mind) for the list of their objectives.  Going through them you should be able to spot the one that should not be there.

**Comment by Profakos** [comment]
"Steal blue documents" is part of this problem.

**Comment by Profakos** [comment]
Okay, I looked over this thing thoroughly. 
It is caused by this line:  init_subtypes(/datum/objective_item,possible_items)
Since objevtive_item's base object, stealabikehorn has this type, it is added. Looks like the initsubtype adds the object with the exact path name too.
The documents show up, because their type is datum/objective_item/unique, which is a subtype of objective_item. 
datum/objective_item/stack/cardboard is not actually a base class of the other stack ores, so they don't inherit the antag restriction list. This line should be added 
datum/objective_item/stack
antag_types = list("Space Ninja")
instead of defining it on the cardboards, just like on special_items.
Other weirdness: special items never appear as objectives for not ninjas, however, neither do stacks other than cardboards.

**Comment by SuperSayu** [comment]
Except that init_subtype deliberately does not do that:
https://github.com/SuperSayu/-tg-station/blob/sayustation/code/__HELPERS/global_lists.dm#L48

> if(path == prototype) continue

prototype is the typepath originally passed in, namely /datum/objective_item, so that object type specifically should not be included in the list.

I do not see a stack/cardboard subtype in objective_items.dm ; instead the cardboard steal is datum/objective_item/stack:
https://github.com/SuperSayu/-tg-station/blob/sayustation/code/game/gamemodes/objective_items.dm#L406

If you could point me at a line and file that says differently it would help.  (If you click a line number, the page url will change to include it, making a jump link as above)

**Comment by Profakos** [comment]
Looking through things closer, then.
https://github.com/SuperSayu/-tg-station/blob/sayustation/code/game/gamemodes/objective_items.dm#L11 I think I found the actual reason why some objectives appear when they didn't have to. This is inherited by the Document Theft objectives too, which should be blocked by all, as it can only be assigned through a special proc.

https://github.com/SuperSayu/-tg-station/blob/sayustation/code/game/gamemodes/objective_items.dm#L367 this line defines that antags for special objectives. None of these appear in our objectives as an error. It is probable that the stack subtypes should be defined like that.

**Comment by Profakos** [comment]
I am sorry I was just a bit worked up as it appears in 90% of the rounds and it has been rather annoying to fix at the start of every round.

**Comment by SuperSayu** [comment]
You don't need to apologize for trying to fix bugs.  It has always been and will always be a noble endeavor.  I am happy to clarify how code works and offer suggestions and interpretations if you need it, I just wish I could spend the time to either hash it out myself or go through it more deeply with you.

**Comment by Profakos** [comment]
Added a debug message after this line https://github.com/SuperSayu/-tg-station/blob/sayustation/code/game/gamemodes/objective.dm#L278

init_subtypes(/datum/objective_item,possible_items)
for(var/datum/objective_item/pitem in possible_items)
    world << "[pitem.name]"

Printed objectives in possible_items:

a red telephone
the captain's antique laser gun
a hand teleporter
a rapid-construction-device
a jetpack
the chief engineer's advanced magnetic boots
a piece of corgi meat
the hypospray
the nuclear authentication disk
an ablative armor vest
the reactive teleport armor
a set of secret documents
the head of security's dermal armor patch
an alien facehugger (dead or alive)
an AI core construction circuit board
an AI upload circuit board
a cyborg upload circuit board
28 moles of plasma (full tank)
a functional AI
the station blueprints
an unused sample of slime extract
four unique identification cards
50 units of polytrinic acid
seven different types of alcohol
four unique blood samples
50 units of unstable mutagen
A silly bike horn! Honk!
the "Red" secret documents
the "Blue" secret documents
A silly bike horn! Honk!
a R.O.B.U.S.T. DELUXE cartridge (Head of Security's PDA cartridge)
a Captain's jumpsuit
a Head of Personnel's jumpsuit
a Head of Security's jumpsuit
a Chief Engineer's jumpsuit
a Research Director's jumpsuit
a Chief Medical Officer's jumpsuit
a Research Director's jumpsuit
a Value-PAK cartridge (Captain's PDA cartridge)
a HumanResources9001 cartridge (Head of Personnel's PDA cartridge)
a R.O.B.U.S.T. DELUXE cartridge (Head of Security's PDA cartridge)
a Power-On DELUXE cartridge (Chief Engineer's PDA cartridge)
a Signal Ace DELUXE cartridge (Research Director's PDA cartridge)
a Med-U DELUXE cartridge (Chief Medical Officer's PDA cartridge)
a Signal Ace DELUXE cartridge (Research Director's PDA cartridge)
a Captain's rubber stamp
a Head of Personnel's rubber stamp
a Head of Security's rubber stamp
a Chief Engineer's rubber stamp
a Research Director's rubber stamp
a Chief Medical Officer's rubber stamp
a Chief Medical Officer's rubber stamp
a Chief Engineer's rubber stamp
the captain's pinpointer
an advanced energy gun
a diamond drill
a bag of holding
a hyper-capacity cell
a laser pointer
a telecom hub circuit board
A silly bike horn! Honk!
10 diamonds
50 gold bars
25 refined uranium bars
5 cardboards

Silly bike horn appears three times. Looks like init_subtypes is not working as intended.
Adding the whole list also adds the 23 head items, but they have a ban on them in the list properly so it just counts as one.

**Comment by SuperSayu** [comment]
First off, you should probably pastebin a long doc like that and link it.

That's a decent debug technique, I would have it output the type as well which will indicate what is giving us problems.

Since you pointed out the cosplay, yes I made a mistake with that--only cosplay/random should have an antag type, I guess, none of the rest should be selected.  Also, the base cosplay type has no name and so might show up as a bike horn... But probably not since it gets renamed in New().

I dunno, will look at it more later.

**Comment by Profakos** [comment]
I looked over it and cosplay except for random is disabled, so it was me overlooking things.
Type display results: http://pastebin.com/6UB8E6kM
It adds 
Name: A silly bike horn! Honk!, type: /datum/objective_item/special
Name: A silly bike horn! Honk!, type: /datum/objective_item/steal
Name: A silly bike horn! Honk!, type: /datum/objective_item/unique
These types are only defined implicitly. They are a valid subtype of /datum/objective_item/steal, therefore, added. Unique is added too.
I suppose this is why on TG possible_items is made by joining multiple lists instead, to avoid this specifically, and also to avoid adding the /datum/objective_item/uniqe, which is not meant to be ever gained through the proc adding random objectives (only example is the exchange of the blue and red documents)

**Comment by Profakos** [comment]
Making progress on this one. Rearranging some item antag types too, based on community feedback.
Bad ones no longer appear. I have also found out that the current cut mechanic skips the last objective.
